File: 1286.Iterator for Combination/backtrace.cpp
class CombinationIterator {
private:
    int index = 0;
    vector<string> combinations;
    void backtrace(const string& characters, int combinationLength, int start, vector<char>& visited){
        if(visited.size() == combinationLength){
            combinations.push_back(string(visited.begin(), visited.end()));
            return;
        }
        int len = characters.size();
        for(int i = start; i < len; ++i){
            char cur = characters[i];
            if(find(visited.begin(), visited.end(), cur) == visited.end()){
                visited.push_back(cur);
                backtrace(characters, combinationLength, i + 1, visited);
                visited.pop_back();
            }
        }
    }
public:
    CombinationIterator(string characters, int combinationLength) {
        vector<char> visited = vector<char>();
        backtrace(characters, combinationLength, 0, visited);
        // copy(combinations.begin(),combinations.end(),ostream_iterator<string>(cout," "));
        // cout << endl;
    }
    
    string next() {
        if(hasNext())
            return combinations[index++];
        assert(false);
    }
    
    bool hasNext() {
        if(index < combinations.size())
            return true;
        return false;
    }
};

/**
 * Your CombinationIterator object will be instantiated and called as such:
 * CombinationIterator* obj = new CombinationIterator(characters, combinationLength);
 * string param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: 1286.Iterator for Combination/backtrace.scala
class CombinationIterator(_characters: String, _combinationLength: Int) {
    import scala.collection.mutable.ArrayBuffer;
    private var index: Int = 0;
    private var combinations: ArrayBuffer[String] = ArrayBuffer[String]();
    private var visited: ArrayBuffer[Char] = ArrayBuffer[Char]();
    
    backtrace(_characters, _combinationLength, 0, visited);
    
    
    def backtrace(characters: String, combinationLength: Int, start: Int, visited: ArrayBuffer[Char]){
        if(visited.length == combinationLength){
            combinations.append(visited.mkString(""));
            return;
        }
        var i = 0;
        var len = characters.length();
        for(i <- start until len){
            var cur: Char = characters.charAt(i);
            if(!visited.contains(cur)){
                visited.append(cur);
                backtrace(characters, combinationLength, i + 1, visited);
                visited.remove(visited.length() - 1);
            }
        }
    }

    def next(): String = {
        if(hasNext()){
            var old: Int = index;
            index += 1;
            return combinations(old);
        }
        assert(false);
        return "";
    }

    def hasNext(): Boolean = {
        if(index < combinations.length)
            return true;
        return false;
    }

}

/**
 * Your CombinationIterator object will be instantiated and called as such:
 * var obj = new CombinationIterator(characters, combinationLength)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */